pkg/imgpkg/registry/registry.go

```diff
@@ -149,42 +167,87 @@ func (r SimpleRegistry) CloneWithSingleAuth(imageRef regname.Tag) (Registry, err
 	keychain := auth.NewSingleAuthKeychain(imgAuth)
 
 	return &SimpleRegistry{
-		remoteOpts: r.remoteOpts,
-		refOpts:    r.refOpts,
-		keychain:   keychain,
+		remoteOpts:      r.remoteOpts,
+		refOpts:         r.refOpts,
+		keychain:        keychain,
+		roundTrippers:   r.roundTrippers,
+		transportAccess: &sync.Mutex{},
 	}, nil
 }
 
-// opts Returns the opts + the keychain
-func (r SimpleRegistry) opts() []regremote.Option {
-	// new options come first so that user configured options take precedence
-	return append([]regremote.Option{regremote.WithAuthFromKeychain(r.keychain)}, r.remoteOpts...)
+// readOpts Returns the readOpts + the keychain
+func (r *SimpleRegistry) readOpts(ref regname.Reference) ([]regremote.Option, error) {
+	rt, err := r.transport(ref, ref.Scope(transport.PullScope))
+	if err != nil {
+		return nil, err
+	}
+	return append([]regremote.Option{regremote.WithAuthFromKeychain(r.keychain), regremote.WithTransport(rt)}, r.remoteOpts...), nil
+}
+
+// writeOpts Returns the writeOpts + the keychain
+func (r *SimpleRegistry) writeOpts(ref regname.Reference) ([]regremote.Option, error) {
+	rt, err := r.transport(ref, ref.Scope(transport.PushScope))
+	if err != nil {
+		return nil, err
+	}
+
+	return append([]regremote.Option{regremote.WithAuthFromKeychain(r.keychain), regremote.WithTransport(rt)}, r.remoteOpts...), nil
+}
+
+// transport Retrieve the RoundTripper that can be used to access the repository
+func (r *SimpleRegistry) transport(ref regname.Reference, scope string) (http.RoundTripper, error) {
+	// The idea is that we can only retrieve 1 RoundTripper at a time to ensure that we do not create
+	// the same RoundTripper multiple times
+	r.transportAccess.Lock()
+	defer r.transportAccess.Unlock()
+	rt := r.roundTrippers.RoundTripper(ref.Context().RegistryStr(), ref.Context().RepositoryStr(), scope)
+	if rt == nil {
+		resolvedAuth, err := r.keychain.Resolve(ref.Context())
+		if err != nil {
+			return nil, fmt.Errorf("Unable retrieve credentials for registry: %s", err)
+		}
+		rt, err = r.roundTrippers.CreateRoundTripper(ref.Context().Registry, resolvedAuth, scope)
```

Comment:
Before doing any request to the registry ggcr will `ping` the registry to understand if it still has access to it, if not it will try and go through the oauth flow again.
You can follow the interaction starting on the Get function
Calling [makeFetcher](https://github.com/google/go-containerregistry/blob/main/pkg/v1/remote/descriptor.go#L219)
Calls [transport.NewWithContext](https://github.com/google/go-containerregistry/blob/main/pkg/v1/remote/transport/transport.go#L38)
Which in turn calls [ping](https://github.com/google/go-containerregistry/blob/main/pkg/v1/remote/transport/ping.go#L71)